eoe/man/man1/ulimit.1
'\"macro stdmacro
.if n .pH g1.ulimit @(#)ulimit      30.2 of 12/25/85
.nr X
.if \nX=0 .ds x} ulimit 1 "Essential Utilities" "\&"
.if \nX=1 .ds x} ulimit 1 "Essential Utilities"
.if \nX=2 .ds x} ulimit 1 "" "\&"
.if \nX=3 .ds x} ulimit "" "" "\&"
.ds OK [\|
.ds CK \|]
.TH \*(x}
.SH NAME
\f3ulimit\f1 - set or report file size limit
.SH SYNOPSIS
\f3ulimit\f1 [-f][blocks]
.SH DESCRIPTION
The \f3ulimit\f1 utility sets or reports the file-size writing limit imposed
on files written by the shell and its child processes (files of any
size may be read). Only a process with appropriate privileges can
increase the limit.
.SH OPTIONS
The \f3ulimit\f1 utility supports the XBD specification, Utility Syntax
Guidelines.
.PP
The following option is supported:
.TP 4
\f3-f\f1
Set (or report, if no blocks operand is present), the file size
limit in blocks. The \f3-f\f1 option is also the default case.
.SH OPERANDS
The following operand is supported:
.TP 8
blocks
The number of 512-byte blocks to use as the new file size limit.
.SH STDOUT
The standard output is used when no blocks operand is present. If the
current number of blocks is limited, the number of blocks in the
current limit is written in the following format:
.IP
"%d\\n", <number of 512-byte blocks>
.PP
If there is no current limit on the number of blocks, in the POSIX
locale the following format is used:
.IP
"unlimited\\n"
.SH EXIT STATUS
The following exit values are returned:
.TP 4
0
Successful completion.
.TP 4
>0
A request for a higher limit was rejected or an error occurred.
.SH APPLICATION USAGE
Since \f3ulimit\f1 affects the current shell execution environment, it is
always provided as a shell regular built-in. If it is called in
separate utility execution environment, such as one of the following:
.IP
.nf
nohup ulimit -f 10000 
env ulimit 10000
.fi
.PP
it will not affect the file size limit of the caller's environment.
.PP
Once a limit has been decreased by a process, it cannot be increased
(unless appropriate privileges are involved), even back to the original
system limit.
.SH EXAMPLES
Set the file size limit to 51,200 bytes:
.IP
ulimit -f 100
.SH SEE ALSO
\f3sh(1), bsh(1)\f1


eoe/cmd/sh/ulimit.c
/*	Copyright (c) 1990, 1991 UNIX System Laboratories, Inc.	*/
/*	Copyright (c) 1984, 1986, 1987, 1988, 1989, 1990 AT&T	*/
/*	  All Rights Reserved  	*/

/*	THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF     	*/
/*	UNIX System Laboratories, Inc.                     	*/
/*	The copyright notice above does not evidence any   	*/
/*	actual or intended publication of such source code.	*/

#ident	"@(#)sh:ulimit.c	1.5.14.1"

/*
 * ulimit builtin
 */

#include <sys/resource.h>
#include <stdlib.h>
#include <pfmt.h>
#include "defs.h"
#ifdef sgi
#include <msgs/uxsgicore.h>
#endif

extern	int	getrlimit64();
extern	int	setrlimit64();

/*
 * order is important in this table! it is indexed by resource ID.
 */

static struct rlimtab {
	char	*name;
	char	*nameid;
	char	*scale;
	char	*scaleid;
	int	divisor;
	ulong	mask;
} rlimtab[] = {
/* RLIMIT_CPU	*/	"time",		":546",	"seconds",	":547",	1,	0x80000000,
/* RLIMIT_FSIZE */	"file",		":548",	"blocks",	":549",	512,	0x80000000,
/* RLIMIT_DATA	*/	"data",		":550",	"kbytes",	":551",	1024,	0xffc00000,
/* RLIMIT_STACK */	"stack",	":552",	"kbytes",	":551",	1024,	0xffc00000,
/* RLIMIT_CORE	*/	"coredump",	":553",	"blocks",	":549",	512,	0x80000000,
/* RLIMIT_NOFILE */	"nofiles",	":554",	"descriptors",	":555",	1,	0x80000000,
/* RLIMIT_VMEM */	"memory",	":556",	"kbytes",	":551",	1024,	0xffc00000,
/* RLIMIT_RSS */	"rss", _SGI_MMX_sh_rss, "kbytes",	":551",	1024,	0xffc00000
};


void
sysulimit(argc, argv)
int argc;
char **argv;
{
	extern int opterr, optind;
	int savopterr, savoptind, savsp;
	char *savoptarg;
	char *args;
	int hard, soft, cnt, c, res;
	rlim64_t limit;
	struct rlimit64 rlimit;
	char resources[RLIM_NLIMITS];

	for (res = 0;  res < RLIM_NLIMITS; res++)
		resources[res] = 0;

	savoptind = optind;
	savopterr = opterr;
	savsp = _sp;
	savoptarg = optarg;
	optind = 1;
	_sp = 1;
	opterr = 0;
	hard = 0;
	soft = 0;
	cnt = 0;

	while ((c = getopt(argc, argv, "HSacdfnstv")) != -1) {
		switch(c) {
			case 'S':
				soft++;
				continue;
			case 'H':
				hard++;
				continue;
			case 'a':
				for (res = 0;  res < RLIM_NLIMITS; res++)
					resources[res]++;
				cnt = RLIM_NLIMITS;
				continue;
			case 'c':
				res = RLIMIT_CORE;
				break;
			case 'd':	
				res = RLIMIT_DATA;
				break;
			case 'f':
				res = RLIMIT_FSIZE;
				break;
			case 'n':	
				res = RLIMIT_NOFILE;
				break;
			case 's':	
				res = RLIMIT_STACK;
				break;
			case 't':	
				res = RLIMIT_CPU;
				break;
			case 'v':	
				res = RLIMIT_VMEM;
				break;
			case 'r':	
				res = RLIMIT_RSS;
				break;
			case '?':
				prusage(SYSULIMIT, ulimuse, ulimuseid);
				goto err;
		}
		resources[res]++;
		cnt++;
	}

	if (cnt == 0) {
		resources[res = RLIMIT_FSIZE]++;
		cnt++;
	}

	/*
	 * if out of arguments, then print the specified resources
	 */

	if (optind == argc) {
		if (!hard && !soft)
			soft++;
		for (res = 0; res < RLIM_NLIMITS; res++) {
			if (resources[res] == 0)
				continue;
			if (res >= sizeof(rlimtab)/sizeof(struct rlimtab))
				break;
			if (getrlimit64(res, &rlimit) < 0)
				continue;
			if (cnt > 1) {
				prs_buff(gettxt(rlimtab[res].nameid, rlimtab[res].name));
				prc_buff('(');
				prs_buff(gettxt(rlimtab[res].scaleid, rlimtab[res].scale));
				prc_buff(')');
				prc_buff(' ');
			}
			if (soft) {
				if (rlimit.rlim_cur == RLIM64_INFINITY)
					prs_buff(gettxt(":557", "unlimited"));
				else 
					prll_buff(rlimit.rlim_cur/rlimtab[res].divisor);
			}
			if (hard && soft)
				prc_buff(':');
			if (hard) {
				if (rlimit.rlim_max == RLIM64_INFINITY)
					prs_buff(gettxt(":557", "unlimited"));
				else 
					prll_buff(rlimit.rlim_max/rlimtab[res].divisor);
			}
			prc_buff('\n');
		}
		goto err;
	}

	if (cnt > 1 || optind + 1 != argc) {
		prusage(SYSULIMIT, ulimuse, ulimuseid);
		goto err;
	}

	if (eq(argv[optind],"unlimited"))
		limit = RLIM64_INFINITY;
	else {
		args = argv[optind];

		limit = 0;
		do {
			if (*args < '0' || *args > '9') {
				error_fail(SYSULIMIT, badulimit, badulimitid);
				goto err;
			}
			limit = (limit * 10) + (*args - '0');
		} while (*++args);

 		limit *= rlimtab[res].divisor;
	}

	if (getrlimit64(res, &rlimit) < 0) {
		error_fail(SYSULIMIT, badulimit, badulimitid);
		goto err;
	}

	if (!hard && !soft)
		hard++, soft++;
	if (hard)
		rlimit.rlim_max = limit;
	if (soft)
		rlimit.rlim_cur = limit;

	if (setrlimit64(res, &rlimit) < 0)
		error_fail(SYSULIMIT, badulimit, badulimitid);

err:
	optind = savoptind;
	opterr = savopterr;
	_sp = savsp;
	optarg = savoptarg;
}
